File: threadsflow/remote_signer.py
"""Signing client for Threads mobile protocol requests."""
from __future__ import annotations
import httpx
from typing import Dict, Any, Optional
from .errors import AuthError, ServerError
from .devices import get_threads_device

class RemoteSigner:
    """Dispatches cryptographic signing requests to the local or cloud signing daemon."""

    def __init__(
        self,
        signing_server: str = "http://127.0.0.1:8643",
        api_key: str = "",
        device_preset: str = "threads_ios",
        timeout: float = 10.0,
    ):
        self.signing_server = signing_server.rstrip("/")
        self.api_key = api_key
        self.device_preset = device_preset
        self.device = get_threads_device(device_preset)
        self._http = httpx.Client(timeout=timeout)

    def sign_request(
        self,
        endpoint: str,
        method: str = "POST",
        body: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        """Sign request through daemon or fallback to authentic local headers."""
        payload = {
            "api_key": self.api_key,
            "device_preset": self.device_preset,
            "endpoint": endpoint,
            "method": method.upper(),
            "platform": "threads",
            "body": body or {},
        }
        try:
            resp = self._http.post(f"{self.signing_server}/v1/sign", json=payload)
            if resp.status_code in (401, 403):
                raise AuthError("Threads signing authorization failed. Valid API key required.")
            if resp.status_code >= 500:
                raise ServerError(f"Signing daemon error: {resp.status_code}")
            return resp.json()
        except (AuthError, ServerError):
            raise
        except Exception:
            # Standalone fallback mode
            return self._fallback_local_sign(endpoint, method, body)

    def _fallback_local_sign(
        self,
        endpoint: str,
        method: str,
        body: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        headers = {
            "User-Agent": self.device.user_agent,
            "X-IG-App-ID": self.device.app_id,
            "X-Bloks-Version-Id": self.device.bloks_version_id,
            "X-IG-Capabilities": self.device.capabilities,
            "X-IG-Connection-Type": "WIFI",
            "Accept-Language": "en-US,en;q=0.9",
            "Content-Type": "application/x-www-form-urlencoded; charset=UTF-8",
        }
        return {"headers": headers, "signed_body": body or {}}

    def close(self):
        self._http.close()

File: ts/src/remote-signer.ts
import { ThreadsDevicePreset } from "./types";
import { getThreadsDevice } from "./devices";
import { AuthError, ServerError } from "./errors";

export interface SignedRequestResult {
  headers: Record<string, string>;
  signed_body?: Record<string, any>;
}

export class RemoteSigner {
  public signingServer: string;
  public apiKey: string;
  public devicePreset: string;
  public device: ThreadsDevicePreset;

  constructor(
    signingServer: string = "http://127.0.0.1:8643",
    apiKey: string = "",
    devicePreset: string = "threads_ios",
  ) {
    this.signingServer = signingServer.replace(/\/+$/, "");
    this.apiKey = apiKey;
    this.devicePreset = devicePreset;
    this.device = getThreadsDevice(devicePreset);
  }

  public async signRequest(
    endpoint: string,
    method: string = "POST",
    body?: Record<string, any>,
  ): Promise<SignedRequestResult> {
    const payload = {
      api_key: this.apiKey,
      device_preset: this.devicePreset,
      endpoint,
      method: method.toUpperCase(),
      platform: "threads",
      body: body || {},
    };

    try {
      const resp = await fetch(`${this.signingServer}/v1/sign`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(payload),
      });

      if (resp.status === 401 || resp.status === 403) {
        throw new AuthError("Threads signing authorization failed. Valid API key required.");
      }
      if (resp.status >= 500) {
        throw new ServerError(`Signing daemon error: ${resp.status}`, resp.status);
      }
      return (await resp.json()) as SignedRequestResult;
    } catch (err: any) {
      if (err instanceof AuthError || err instanceof ServerError) {
        throw err;
      }
      return this.fallbackLocalSign(endpoint, method, body);
    }
  }

  public fallbackLocalSign(
    _endpoint: string,
    _method: string,
    body?: Record<string, any>,
  ): SignedRequestResult {
    const headers: Record<string, string> = {
      "User-Agent": this.device.userAgent,
      "X-IG-App-ID": this.device.appId,
      "X-Bloks-Version-Id": this.device.bloksVersionId,
      "X-IG-Capabilities": this.device.capabilities,
      "X-IG-Connection-Type": "WIFI",
      "Accept-Language": "en-US,en;q=0.9",
      "Content-Type": "application/x-www-form-urlencoded; charset=UTF-8",
    };

    return {
      headers,
      signed_body: body || {},
    };
  }
}
